### PR TITLE
Adding test run attachments processing

### PR DESCRIPTION
**Pull request by jakubch1**
## Description
This PR is adding multi test run finalization process which can be used to reprocess/merge attachments coming from separated test run executions. This logic will be mainly used to merge code coverage reports from multi test runs.

## Related issue
#1811

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                    {
+                        await task;
+                        eventHandler?.HandleMultiTestRunFinalizationComplete(attachments);
+                        testPlatformEventSource.MultiTestRunFinalizationStop(attachments.Count);
+                        return attachments;
+                    }
+                    else
+                    {
+                        eventHandler?.HandleLogMessage(ObjectModel.Logging.TestMessageLevel.Informational, "Finalization was cancelled.");
+                        eventHandler?.HandleMultiTestRunFinalizationComplete(null);
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs:75]: >eventHandler?.HandleMultiTestRunFinalizationComplete(null); [](start = 24, length = 59)

Can we send the current set of attachments in this case?

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs [hunk 2]
```diff
@@ -725,6 +731,61 @@ private async Task SendMessageAndListenAndReportTestResultsAsync(string messageT
...
+                            var multiTestRunFinalizationCompletePayload = this.dataSerializer.DeserializePayload<MultiTestRunFinalizationCompletePayload>(message);
+
+                            eventHandler.HandleMultiTestRunFinalizationComplete(multiTestRunFinalizationCompletePayload.Attachments);
+                            isMultiTestRunFinalizationComplete = true;
+                        }
+                        else if (string.Equals(MessageType.TestMessage, message.MessageType))
+                        {
+                            var testMessagePayload = this.dataSerializer.DeserializePayload<TestMessagePayload>(message);
+                            eventHandler.HandleLogMessage(testMessagePayload.MessageLevel, testMessagePayload.Message);
+                        }
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs:770]: > } [](start = 23, length = 2)

Should we log if we get a message we cannot recognize?

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs [hunk 3]
```diff
@@ -725,6 +731,61 @@ private async Task SendMessageAndListenAndReportTestResultsAsync(string messageT
...
+                eventHandler.HandleLogMessage(TestMessageLevel.Error, TranslationLayerResources.AbortedMultiTestRunFinalization);               
+                eventHandler.HandleMultiTestRunFinalizationComplete(null);
+
+                // Earlier we were closing the connection with vstest.console in case of exceptions
+                // Removing that code because vstest.console might be in a healthy state and letting the client
+                // know of the error, so that the TL can wait for the next instruction from the client itself.
+                // Also, connection termination might not kill the process which could result in files being locked by testhost.
+            }
+            
+            this.testPlatformEventSource.TranslationLayerMultiTestRunFinalizationStop();
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs:786]: >this.testPlatformEventSource.TranslationLayerMultiTestRunFinalizationStop();  [](start = 12, length = 77)

should this be in a finally block?

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs [hunk 4]
```diff
@@ -119,6 +120,14 @@ public interface IVsTestConsoleWrapper
         /// <param name="customTestHostLauncher">Custom test host launcher for the run.</param>
         void RunTestsWithCustomTestHost(IEnumerable<TestCase> testCases, string runSettings, TestPlatformOptions options, ITestRunEventsHandler testRunEventsHandler, ITestHostLauncher customTestHostLauncher);
 
+        /// <summary>
+        /// Provides back all attachments to TestPlatform for additional processing (for example merging)
+        /// </summary>
+        /// <param name="attachments">Collection of attachments</param>
+        /// <param name="multiTestRunFinalizationCompleteEventsHandler">EventHandler to receive session complete event</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        Task FinalizeMultiTestRunAsync(ICollection<AttachmentSet> attachments, IMultiTestRunFinalizationEventsHandler multiTestRunFinalizationCompleteEventsHandler, CancellationToken cancellationToken);
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs:129]: Tagging @nohwnd  and @cvpoienaru on the changes in this file and its async equivalent. The assumption is that external consumers do not implement this interface so its ok to change it without introducing a IVSTestConsoleWrapper2. Do you have any concerns?

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/Resources.Designer.cs [hunk 5]
```diff
@@ -52,55 +52,64 @@ public class Resources {
...
                 return ResourceManager.GetString("AbortedTestsDiscovery", resourceCulture);
             }
         }
-
+        
         /// <summary>
         ///   Looks up a localized string similar to The active Tests Run was aborted..
         /// </summary>
-        public static string AbortedTestsRun {
+        internal static string AbortedTestsRun {
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/Resources.Designer.cs:85]: >internal [](start = 8, length = 8)

Was this intentionally made internal? I'm not sure if this was meant to be used by external folks.

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs (hunk 4 above)
Comment (nohwnd) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs:129]: I don't have any. All usages of this interface I saw were consuming it, not implementing it.

Comment (jakubch1) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs:129]: >IMultiTestRunFinalizationEventsHandler [](start = 79, length = 38)

Add telemetry bool and IEnumerable. Recheck handler to be in same project as handlers for test run

File: src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/IMultiTestRunFinalizationEventsHandler.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    /// <summary>
+    /// Interface contract for handling multi test run finalization complete events
+    /// </summary>
+    public interface IMultiTestRunFinalizationEventsHandler : ITestMessageEventHandler
+    {
+        /// <summary>
+        /// Dispatch MultiTestRunFinalizationComplete event to listeners.
+        /// </summary>
+        /// <param name="attachments">Attachments reprocessed.</param>
+        void HandleMultiTestRunFinalizationComplete(ICollection<AttachmentSet> attachments);
```
Comment (jakubch1) [inline src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/IMultiTestRunFinalizationEventsHandler.cs:17]: Add progress callback

current number of processor / number of all processors

(current attachments index / number of all attachments for current processor) -> number percentage

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/Resources.Designer.cs (hunk 5 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/Resources.Designer.cs:85]: this file is automatically generated. All other stuff is also internal

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs (hunk 4 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs:129]: added

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs (hunk 3 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs]: fixed

File: src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs (hunk 1 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs]: fixed

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs (hunk 2 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs:772]: we can but this is not current pattern. Discovery and test execution are implemented in same way

Comment (AbhitejJohn) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs:772]: I understand, perhaps we could fix that for the new code we add in. Doesn't seem like a risky change and would help for troubleshooting. Ideally we should be throwing.

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/Resources.Designer.cs (hunk 5 above)
Comment (AbhitejJohn) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/Resources.Designer.cs:85]: Oh the access modifiers for a resource file can be updated in VS. It's the strip just above all the strings. Looks like for some reason that got updated to internal in this PR (perhaps adding a new string causes that).
image

File: src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/IMultiTestRunFinalizationEventsHandler.cs (hunk 6 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.ObjectModel/Client/Interfaces/IMultiTestRunFinalizationEventsHandler.cs]: added

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/Resources.Designer.cs (hunk 5 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/Resources.Designer.cs]: fixed, it looks like wrong value was inside csproj

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs (hunk 4 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapper.cs]: I didn't find any implementations of it. Resolving this.

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs (hunk 2 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs:777]: Looking how other stuff is implemented I think throwing in this case is dangerous. I think still we can get some late messages from test run or discovery. I prefer log error in this case and send this error back through EventsHandler. WDYT?

Comment (AbhitejJohn) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs:777]: yup, logging was what I was recommending too.

File: src/Microsoft.TestPlatform.Client/MultiTestRunFinalization/MultiTestRunFinalizationEventsHandler.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                FinalizationProgressEventArgs = finalizationProgressEventArgs,
+            };
+
+            this.communicationManager.SendMessage(MessageType.MultiTestRunFinalizationProgress, payload);
+        }
+
+        /// <inheritdoc/>
+        public void HandleProcessedAttachmentsChunk(IEnumerable<AttachmentSet> attachments)
+        {
+            throw new System.NotImplementedException();
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.Client/MultiTestRunFinalization/MultiTestRunFinalizationEventsHandler.cs:60]: >NotImplementedException [](start = 29, length = 23)

Is this a TODO?

Comment (jakubch1) [reply src/Microsoft.TestPlatform.Client/MultiTestRunFinalization/MultiTestRunFinalizationEventsHandler.cs:60]: We are not using this at all currently. It is for future when we will have more attachment processors

File: src/Microsoft.TestPlatform.Common/Interfaces/Engine/IMultiTestRunFinalizationManager.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        /// <param name="cancellationToken">Cancellation token</param>
+        Task FinalizeMultiTestRunAsync(IRequestData requestData, IEnumerable<AttachmentSet> attachments, IMultiTestRunFinalizationEventsHandler eventHandler, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Finalizes multi test run
+        /// </summary>
+        /// <param name="attachments">Collection of attachments</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Collection of attachments.</returns>
+        Task<Collection<AttachmentSet>> FinalizeMultiTestRunAsync(IRequestData requestData, IEnumerable<AttachmentSet> attachments, CancellationToken cancellationToken);
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.Common/Interfaces/Engine/IMultiTestRunFinalizationManager.cs:31]: >FinalizeMultiTestRunAsync [](start = 40, length = 25)

Do we really need this overload? Seems like the callers can pass in null.

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs (hunk 2 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/VsTestConsoleRequestSender.cs:777]: warning added

File: src/Microsoft.TestPlatform.Common/Interfaces/Engine/IMultiTestRunFinalizationManager.cs (hunk 8 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.Common/Interfaces/Engine/IMultiTestRunFinalizationManager.cs:31]: There is also difference that one method is returning attachments while second is not returning but calling handler.

File: src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+            var logger = CreateMessageLogger(eventsHandler);
+
+            for (int i = 0; i < dataCollectorAttachmentsProcessors.Length; i++)
+            {
+                var dataCollectorAttachmentsProcessor = dataCollectorAttachmentsProcessors[i];
+                int attachmentsHandlerIndex = i + 1;
+
+                ICollection<Uri> attachementProcessorUris = dataCollectorAttachmentsProcessor.GetExtensionUris()?.ToList();
+                if (attachementProcessorUris != null && attachementProcessorUris.Any())
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs:111]: >attachementProcessorUris [](start = 56, length = 24)

nit: spelling `attachmentProcessorUris`

File: src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+            }
+        }
+
+        private async Task<Collection<AttachmentSet>> ProcessAttachmentsAsync(Collection<AttachmentSet> attachments, IMultiTestRunFinalizationEventsHandler eventsHandler, CancellationToken cancellationToken)
+        {
+            if (attachments == null || !attachments.Any()) return attachments;
+
+            var logger = CreateMessageLogger(eventsHandler);
+
+            for (int i = 0; i < dataCollectorAttachmentsProcessors.Length; i++)
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs:105]: >dataCollectorAttachmentsProcessors.Length [](start = 32, length = 41)

I'm guessing the ordering of attachment processors is a TODO in this PR? Perhaps we could add a note and bring that in another PR.

File: src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs (hunk 9 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs]: fixed

File: src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs (hunk 10 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs:105]: Do we want to expose priority on Attachment Processor interface?

File: src/Microsoft.TestPlatform.ObjectModel/DataCollector/IDataCollectorAttachmentProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        /// <summary>
+        /// Gets the attachments Uris, which are handled by attachment processor
+        /// </summary>
+        IEnumerable<Uri> GetExtensionUris();
+
+        /// <summary>
+        /// Indicates whether attachment processor is supporting incremental processing of attachments
+        /// </summary>
+        bool SupportsIncrementalProcessing { get; }
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.ObjectModel/DataCollector/IDataCollectorAttachmentProcessor.cs:26]: An example with a little more detail in the comments would be very useful.

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Microsoft.TestPlatform.VsTestConsole.TranslationLayer.csproj [hunk 12]
```diff
@@ -37,7 +37,7 @@
   </ItemGroup>
   <ItemGroup>
     <EmbeddedResource Update="Resources\Resources.resx">
-      <Generator>ResXFileCodeGenerator</Generator>
+      <Generator>PublicResXFileCodeGenerator</Generator>
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Microsoft.TestPlatform.VsTestConsole.TranslationLayer.csproj:40]: Is this change intentional?

Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Microsoft.TestPlatform.VsTestConsole.TranslationLayer.csproj:40]: yes, this is what you asked me to do to not generate something internal, but public

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapperAsync.cs [hunk 13]
```diff
@@ -82,18 +78,13 @@ public interface IVsTestConsoleWrapperAsync
         Task RunTestsWithCustomTestHostAsync(IEnumerable<TestCase> testCases, string runSettings, TestPlatformOptions options, ITestRunEventsHandler testRunEventsHandler, ITestHostLauncher customTestHostLauncher);
 
         /// <summary>
-        /// See <see cref="IVsTestConsoleWrapper.CancelTestRun"/>.
-        /// </summary>
-        void CancelTestRun();
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapperAsync.cs:87]: >CancelTestRun [](start = 13, length = 13)

Is there any way we could keep this by hiding it in the derived interface?

Comment (AbhitejJohn) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapperAsync.cs:87]: Removing APIs would definitely break people.

---
In reply to: [448171317](https://github.com/microsoft/vstest/pull/2463#discussion_r448171317) [](ancestors = 448171317)

File: src/Microsoft.TestPlatform.ObjectModel/DataCollector/IDataCollectorAttachmentProcessor.cs (hunk 11 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.ObjectModel/DataCollector/IDataCollectorAttachmentProcessor.cs:38]: added same explanation like in RFC

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/xlf/Resources.de.xlf [hunk 14]
```diff
@@ -61,6 +61,11 @@
         <target state="translated">Die Datei "{0}" ist nicht vorhanden.</target>
         <note></note>
       </trans-unit>
+      <trans-unit id="AbortedMultiTestRunFinalization">
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Resources/xlf/Resources.de.xlf:64]: >AbortedMultiTestRunFinalization [](start = 22, length = 31)

You'd probably want to ask for a special localization test pass for this or ask for a loc exception.

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs [hunk 15]
```diff
@@ -301,6 +312,45 @@ public void RunTests(TestRunRequestPayload testRunRequestPayload, ITestHostLaunc
             }
         }
 
+        /// <inheritdoc/>
+        public void FinalizeMultiTestRun(MultiTestRunFinalizationPayload finalizationPayload, IMultiTestRunFinalizationEventsHandler finalizationEventsHandler, ProtocolConfig protocolConfig)
```
Comment (AbhitejJohn) [inline src/vstest.console/TestPlatformHelpers/TestRequestManager.cs:316]: Perhaps a similar natured question was asked in the RFC as well, but any reason this method is not async?

Comment (AbhitejJohn) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs:316]: You wouldn't need a separate Cancel API below in that case.

---
In reply to: [448174262](https://github.com/microsoft/vstest/pull/2463#discussion_r448174262) [](ancestors = 448174262)

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/MultiTestRunFinalization/MultiTestRunFinalizationManagerTests.cs [hunk 16]
```diff
@@ -0,0 +1,625 @@
...
+            mockEventsHandler.Verify(h => h.HandleMultiTestRunFinalizationProgress(It.Is<MultiTestRunFinalizationProgressEventArgs>(a => VerifyProgressArgs(a, 3))));
+            mockEventsHandler.Verify(h => h.HandleMultiTestRunFinalizationProgress(It.Is<MultiTestRunFinalizationProgressEventArgs>(a => VerifyProgressArgs(a, 4))));
+            mockEventsHandler.Verify(h => h.HandleLogMessage(TestMessageLevel.Informational, "Finalization was cancelled."));
+            mockAttachmentHandler1.Verify(h => h.GetExtensionUris());
+            mockAttachmentHandler2.Verify(h => h.GetExtensionUris(), Times.Never);
+            mockAttachmentHandler1.Verify(h => h.ProcessAttachmentSetsAsync(It.Is<ICollection<AttachmentSet>>(c => c.Count == 1 && c.Contains(inputAttachments[0])), It.IsAny<IProgress<int>>(), It.IsAny<IMessageLogger>(), cancellationTokenSource.Token));
+            mockAttachmentHandler2.Verify(h => h.ProcessAttachmentSetsAsync(It.IsAny<ICollection<AttachmentSet>>(), It.IsAny<IProgress<int>>(), It.IsAny<IMessageLogger>(), It.IsAny<CancellationToken>()), Times.Never);
+
+            VerifyMetrics(1, 1, "Canceled");
```
Comment (AbhitejJohn) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/MultiTestRunFinalization/MultiTestRunFinalizationManagerTests.cs:453]: >1 [](start = 26, length = 1)

consider using named parameters for all Verify calls.

**Review by AbhitejJohn — APPROVED** [review]
:shipit:

**Comment by AbhitejJohn** [comment]
When both this and the VSUnitTesting side of the changes are ready. please create a merged insertion PR and send that out for P0/P1 scenario validation to the vendors. Feel free to include specific workflows you'd want them to validate as well.

File: src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapperAsync.cs (hunk 13 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.VsTestConsole.TranslationLayer/Interfaces/IVsTestConsoleWrapperAsync.cs]: fixed

File: src/vstest.console/TestPlatformHelpers/TestRequestManager.cs (hunk 15 above)
Comment (jakubch1) [reply src/vstest.console/TestPlatformHelpers/TestRequestManager.cs]: In RFC it was different place in code which was already fixed. Here we cant use async method as we can't use await inside lock. It also would not help with canceling as this would mean that DesignClient would need to keep cancellationtoken. I would be consistent here.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/MultiTestRunFinalization/MultiTestRunFinalizationManagerTests.cs (hunk 16 above)
Comment (jakubch1) [reply test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/MultiTestRunFinalization/MultiTestRunFinalizationManagerTests.cs]: fixed

File: src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs (hunk 10 above)
Comment (jakubch1) [reply src/Microsoft.TestPlatform.CrossPlatEngine/MultiTestRunFinalization/MultiTestRunFinalizationManager.cs]: We can add it later as attribute. Resolving

File: src/Microsoft.TestPlatform.ObjectModel/DataCollector/IDataCollectorAttachmentProcessor.cs [hunk 17]
```diff
@@ -22,6 +22,18 @@ public interface IDataCollectorAttachmentProcessor
 
         /// <summary>
         /// Indicates whether attachment processor is supporting incremental processing of attachments
+        /// 
+        /// `SupportsIncrementalProcessing` should indicate if attachment processor is supporting incremental processing of attachments. It means that `ProcessAttachmentSetsAsync` should be [associative](https://en.wikipedia.org/wiki/Associative_property).
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.ObjectModel/DataCollector/IDataCollectorAttachmentProcessor.cs:26]: Consider putting detailed explanation in the <remarks> section, so intellisense is manageable.

Comment (jakubch1) [reply src/Microsoft.TestPlatform.ObjectModel/DataCollector/IDataCollectorAttachmentProcessor.cs:26]: fixed
